lib/invoker.js

```diff
@@ -24,21 +25,63 @@ const invoke = (userFunction) => {
     };
 };
 
-module.exports = (userFunctionUri, port) => {
-    const userFunction = lookUpFunction(userFunctionUri);
+const registerShutdown = (destroyHook, timeout, server, shutdownSignal) => {
+    return shutdownSignal.then(() => {
+        logger('Graceful shutdown started');
+        return new Promise(((resolve, reject) => {
+            server.tryShutdown(() => {
+                executeDestroy(destroyHook, timeout)
+                    .then(() => {
+                        logger('Graceful shutdown completed');
+                        resolve();
+                    }, (err) => {
+                        logger(`Error during $destroy hook execution: ${err.toString()}`);
+                        reject(err);
+                    });
+            });
+        }));
+    });
+};
+
+const executeInit = async (initHook, initTimeout) => {
+    logger('Calling $init hook (if any)');
+    return await guardWithTimeout(initHook, initTimeout)
+        .then(() => logger('$init hook successfully completed (or skipped if none)'));
+};
 
+const executeDestroy = async (destroyHook, timeout) => {
+    logger('Calling $destroy hook (if any)');
+    return await guardWithTimeout(destroyHook, timeout)
+        .then(() => logger('$destroy hook successfully completed (or skipped if none)'));
+};
+
+module.exports = async (userFunctionUri, options, shutdownSignal) => {
     const server = new grpc.Server();
-    server.addService(services.RiffService, {invoke: invoke(userFunction)});
 
-    const shutdown = () => {
-        logger('Graceful shutdown started');
-        server.tryShutdown(() => logger('Graceful shutdown completed'));
-    };
-    process.on('SIGTERM', shutdown);
-    process.on('SIGINT', shutdown);
+    const userFunction = lookUpFunction(userFunctionUri);
+
+    const destroyHook = userFunction.$destroy;
+    const destroyHookTimeout = options.$destroyTimeoutMs || 10000;
+    const shutdownPromise = registerShutdown(destroyHook, destroyHookTimeout, server, shutdownSignal);
+
+    const initHookTimeout = options.$initTimeoutMs || 10000;
+    await executeInit(userFunction.$init, initHookTimeout, server)
+        .then(
+            () => {
+                server.addService(services.RiffService, {invoke: invoke(userFunction)});
+                server.bind(`0.0.0.0:${options.port}`, grpc.ServerCredentials.createInsecure());
+                server.start();
+                logger('Ready to process signals');
+            },
+            (err) => {
```

Comment:
the node process is still running after running this errback.

It would be better to trigger the main shutdown flow rather then try to hack in an alternative

Reply:
this is actually a place where unit tests will never be sufficient, the only way to effectively test this behavior is with a clean node process.

Reply:
I have manually run all the cases (combo of sync vs. async hooks, KO vs OK hooks) and the server always shuts downs.

Reply:
The only case where it takes a while (as in: a few seconds on my machine) is when the $destroy hook is synchronous and fails. Otherwise, the shutdown seems instantaneous.

Reply:
Just updated the initial comment with the validation steps I am following.

Reply:
> The only case where it takes a while (as in: a few seconds on my machine) is when the $destroy hook is synchronous and fails.

That's the problem. This means that the node process is waiting for a timer to finish before the process can exit. Since user code can define their down timeouts, we can't wait for all timers to finish before exiting.

Reply:
Now fixed